Feat/migrate to clap

xnacly opened the pull request:
# Migration to clap
fixes: #14 
fixes: #3
improves: #12
lays foundation for: #16 
## Changes:
- migrated to clap:
  ```text
  Usage: opus <COMMAND>
  
  Commands:
    add     create a new task [aliases: a]
    delete  delete a task with the given id [aliases: del, d]
    clear   remove all tasks from the database
    finish  mark the task with the given id as finished [aliases: fin, f]
    list    list tasks matching the given query [aliases: ls, l]
    export  export all tasks
    help    Print this message or the help of the given subcommand(s)
  
  Options:
    -h, --help     Print help information
    -V, --version  Print version information
  ```
- added `--finished` flag to `opus ls`:
  ```text
  list tasks matching the given query
  
  Usage: opus list [OPTIONS] [QUERY]
  
  Arguments:
    [QUERY]  
  
  Options:
    -f, --finished  displays tasks marked as finished
    -h, --help      Print help information
    -V, --version   Print version information
  ```
- reworked `opus export` to provide a better interface:
  - before
    ```bash
    # exports all tasks in the data.json file
    opus export json data
    # export all tasks in the mycsvfile.csv file
    opus export csv mycsvfile
    # data.tsv
    opus export tsv data
    ```
  - after
    ```bash
    # exports all tasks in the data.json file
    opus export --format="json" --output="data.json"
    # export all tasks in the data.csv file
    opus export --format="csv" --output="data.csv"
    # export all tasks in the data.tsv file
    opus export --format="tsv" --output="data.csv"
    ```
- switched from specifing a tasks priority with symbols `,` to numbers prefixed with `.`
  - before:
    ```bash
    # add a new task with the following properties:
    # title: review and merge pr 5
    # due: @tomorrow
    # tag: #github
    # priority: 3
    opus add "review and merge pr 5 @tomorrow #github ,,,"
    ```
  - after:
    ```bash
    # add a new task with the following properties:
    # title: review and merge pr 5
    # due: @tomorrow
    # tag: #github
    # priority: 3
    opus add "review and merge pr 5 @tomorrow #github .3"
    ```

- rewrote tests and removed everything regarding the self made args parser
- improved task displaying:
  - before:
    ```text
    [2]: 'update excel sheet' (2022-10-13) #work [,,,]
    ```
  - after:
    ```
    [2]: 'update excel sheet' (2022-10-13) #work .3 FINISHED
    ```

xnacly committed:
feat: 'list' and '--finished'

migrated opus list to clap logic, implemented displaying finished tasks via the --finished flag

xnacly committed:
feat: close db after main loop end

xnacly committed:
doc: add '--finished' to list usage

xnacly committed:
feat: add Task::from(String)

i got stuff to do #work .5 @2022-10-17

translates to:

[src/tests.rs:291] &t = Task {
    id: None,
    title: "i got stuff to do",
    tag: "#work",
    priority: 5,
    due: "2022-10-17",
    finished: false,
}

- this commit changes the behaviour of specifing the priority by commatas to the new implementation which consists of prefixing the prio with a '.': Priority of 5 is to be written as '.5'
- updated querying by priority in opus list
- add a test for parsing
- update usage

xnacly committed:
tests: rewrote tests to match migration

xnacly committed:
feat: migrate opus add
- panic if task title is empty

xnacly committed:
feat: add aliases to commands

xnacly committed:
feat: add a FINISHED info to a task

- FINISHED information now appears behind a tasks content if its marked as finished
- priority changes are now reflected in printing tasks:

  before:
   [2]: 'update excel sheet' (2022-10-13) #work [,,,]
   after:
   [2]: 'update excel sheet' (2022-10-13) #work .3 FINISHED

xnacly committed:
feat: migrate more commands
- implement all commands
  - clear
  - delete
  - finish
  - export
- remove unused types
- remove old logic
- added changes suggested by clippy
- update usage

xnacly committed:
doc: reflect prio changes in doc comment

xnacly committed:
Merge branch 'master' into feat/migrate-to-clap

xnacly committed:
fix: fix issues caused by github merge

hlxid commented on src/db.rs:
```diff
@@ -82,13 +87,12 @@ impl Database {
                 .map(|x| x.expect("Couldn't map over tasks returned by database"))
-                // TODO: remove this if queried by id
-                .filter(|x| !x.finished)
+                .filter(|x| if display_finished { true } else { !x.finished })
```
nit: since this is a logic expression and not used to manage control flow, I would personally rewrite this using an logical OR
```suggestion
                .filter(|x| display_finished || !x.finished )
```

hlxid commented on src/db.rs:
```diff
@@ -103,8 +107,7 @@ impl Database {
         .expect("Couldn't get task with the given query")
         .map(|x| x.expect("Couldn't map over tasks returned by database"))
-        // TODO: remove
-        .filter(|x| !x.finished)
+        .filter(|x| if display_finished { true } else { !x.finished })
         .collect::<Vec<Task>>()
```
Not specific to the clap migration but I want to mention this nonetheless:
These lines are redundant to the ones found in line 89-91. The lambda that is passed to `query_map` is redundant as well and results in some duplicate code.

xnacly replied:
Sadly i have to violate the DRY principle here, due to the fact that i need one query implementation without params (query all tasks) and one with params (all other queries)

hlxid replied:
Having two separate calls to `query_map` is necessary here, yes. However you can save the result outside the if and then do the mapping, filtering etc outside the if. Additionally you can save the lambda that is passed to `query_map` in a variable to only have to define it once.

hlxid replied:
Something like this should work (may need a import to the SQL row class):
```rust
        let mut stmt = self
            .con
            .prepare(sql_query)
            .expect("Failed to prepare SQL statement in querying for tasks");

        if sql_query == GET_TASK_BY_PRIO {
            query = query[1..].to_string();
        }

        let row_converter = |row: &Row| {
            Ok(Task {
                id: row.get("id")?,
                title: row.get("title")?,
                tag: row.get("tag")?,
                due: row.get("due")?,
                priority: row.get("priority")?,
                finished: matches!(row.get("finished")?, 1),
            })
        };

        let result = if query == "list" || query == "l" {
            stmt.query_map([], row_converter)
        } else {
            stmt.query_map([query], row_converter)
        };

        result
            .expect("Couldn't get task with the given query")
            .map(|x| x.expect("Couldn't map over tasks returned by database"))
            .filter(|x| if display_finished { true } else { !x.finished })
            .collect::<Vec<Task>>()
```

hlxid commented on src/main.rs:
```diff
@@ -18,66 +13,159 @@ mod util;
+                .get_one::<String>("fileformat")
+                .expect("Bad Fileformat");
+
+            let export_type = match temp.as_str() {
+                "csv" => ExportType::Csv,
+                "json" => ExportType::Json,
+                "tsv" => ExportType::Tsv,
+                _ => panic!("Unknown exporttype"),
+            };
```
Since you moved the task parsing to `Task::from` with this pull request I encourage you to do the same for the `ExportType` enum. The conversion to a string is already there using the `Display` trait and imho the conversion in the other direction should as well be implemented by the enum to have both conversions at the same place.

hlxid commented on src/tests.rs:
```diff
@@ -1,279 +1,163 @@
-        assert_eq!(task.title, "update excel sheet");
-        assert_eq!(task.tag, "#work");
-        assert_eq!(task.priority, 3);
+        assert_eq!(task.title, task1.title);
+        assert_eq!(task.tag, task1.tag);
+        assert_eq!(task.priority, task1.priority);
+        assert_eq!(task.due, task1.due);
+        assert_eq!(task.finished, task1.finished);
```
Maybe we could define a function e.g. `Task::equals_content` that checks whether the content (everything except the id) is the same to reduce code redundancy since this is used three times.

Because that function would be only used inside the tests module you would receive a compiler warning that it is unused (since you re-enabled dead code warnings :+1:). To fix this you could place the impl block for that function inside the test module or add `#[cfg(test)]` to the function definition.

hlxid commented on src/tests.rs:
```diff
@@ -1,279 +1,163 @@
+        let id = db.con.last_insert_rowid();
 
-        cli_fin_task(&db, id.clone());
+        cli_fin_task(&db, id.clone().to_string());
```
The clone isn't necessary here when using `to_string()` because it doesn't consume the number.
```suggestion
        cli_fin_task(&db, id.to_string());
```

hlxid commented on src/types.rs:
```diff
@@ -58,6 +32,52 @@ pub struct Task {
     pub finished: bool,
 }
 
+impl From<String> for Task {
```
Is there any special reason for using owned `String`s here instead of `&str` that I am missing?
Generally you would to my knowledge only use owned `String`s when you are saving the original string inside the struct and want to move the ownership to the struct instance.

If the caller has the data as a `&str` or creates multiple task with the same `String` currently the caller would need to make a copy of the string, which ownership is passed to `Task::from` which is then dropped after parsing. This copy can be removed because you are coping the string in the parsing function anyways and do not need the ownership of it.

There is even a special trait for parsing `&str` into a struct: [`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html). However this trait expects you to return a `Result` so you may use `From<&str>` instead because we don't really want to do proper error handling in this cli and panicking is fine in our usecase.

This change would also simplify the tests where you always convert a `&str` into a owned `String` in order to be able to call this function.

hlxid commented on src/types.rs:
```diff
@@ -78,18 +98,11 @@ impl fmt::Display for Task {
+        if self.finished {
+            write!(f, " FINISHED")?;
         }
         write!(f, "")
```
nit: this call seems doesn't really do anything. I think this can be safely removed.

xnacly replied:
It cant, its purely there to indicate if a task is marked as finished, the call in line 106 is there to have the correct return type

hlxid replied:
Ah, yes, you're right. My bad.

hlxid reviewed (changes requested):
Here are a few code quality related things that you may want to look at.
Functionality-wise this looks good to me.

hlxid commented on Cargo.toml:
```diff
@@ -1,10 +1,13 @@
 name = "opus"
-version = "0.1.0-alpha"
+version = "0.2.0-alpha"
+author = "xnacly"
+about = "manage tasks with ease"
```
If you intend to display these on the help page, you'll still need to add them like this: https://docs.rs/clap/latest/clap/macro.crate_authors.html and https://docs.rs/clap/latest/clap/macro.crate_description.html

xnacly committed:
feat: implemented suggestions made by @daniel0611

- replace whole if statement with logical 'or' / '||'
- add ExportType::from(&str)
- update Task::from(&str)
- add Task.compare
- add makro to extract crate description

hlxid reviewed (approved):
LGTM
